File: app/src/main/java/com/example/gb_android_base_appnotes/observe/Publisher.java
package com.example.gb_android_base_appnotes.observe;

import com.example.gb_android_base_appnotes.data.CardNote;

import java.util.ArrayList;
import java.util.List;

public class Publisher {
    private List<Observer> observers;

    public Publisher() {
        observers = new ArrayList<>();
    }

    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }

    public void notifySingle(CardNote cardNote) {
        for (Observer observer : observers) {
            observer.updateCardNote(cardNote);
            unsubscribe(observer);
        }
    }
}

File: app/src/main/java/com/example/gb_android_base_appnotes/observe/Publisher.kt
package com.example.gb_android_base_appnotes.observe

import com.example.gb_android_base_appnotes.data.CardNote
import java.util.*

class Publisher {
    private val observers: MutableList<Observer>
    fun subscribe(observer: Observer) {
        observers.add(observer)
    }

    fun unsubscribe(observer: Observer) {
        observers.remove(observer)
    }

    fun notifySingle(cardNote: CardNote?) {
        for (observer in observers) {
            observer.updateCardNote(cardNote)
            unsubscribe(observer)
        }
    }

    init {
        observers = ArrayList()
    }
}